sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,549 @@
+ executing the task function and if it can write its results to the output data
+ structure _D_, then it does so and some other task thread in _S_ is
+ strengthened to have concurrent forward progress guarantee. If a task thread
+ cannot write its results to _D_, the task thread blocks until space is
+ available.
+
+The two examples below, respectively, show the following:
+
+1. How strengthening of a work item strengthens the task threads.
+
+2. How a task thread delegates its progress guarantee to other task threads in
+the same task sequence object.
+
+Example 1 uses the following program:
+
+```
+// A kernel K
+{
+  ...
+  task_sequence<some_function, ...> task_obj1; // Object_1_Thread
+  task_obj1.async(...); // Object_1_Task_1_Thread
+  task_obj1.async(...); // Object_1_Task_2_Thread
+  ...
+  task_sequence<some_function, ...> task_obj2; // Object_2_Thread
+  task_obj2.async(...); // Object_2_Task_1_Thread
+  task_obj2.async(...); // Object_2_Task_2_Thread
+}
+```
+The call to the task_sequence constructor creates the task object threads
+_Object_1_Thread_ and _Object_2_Thread_. The first two `async` calls create task
+threads _Object_1_Task_1_Thread_ and _Object_1_Task_2_Thread_. Similarly the
+next two calls create _Object_2_Task_1_Thread_ and _Object_2_Task_2_Thread_.
+
+The table below provides a view of the hierarchy of task threads that will be
+generated.
+
+.Hierarchy of task threads.
+[cols="s,,,,"]
+|=====
+// row 1, cells 2 spans 4 cells hence the '4+' before '|'
+| Work Item 4+^| _WI_
+// row 2, cells after the first one span 2 cells each
+|Task Sequence Object Thread
+2+^|_Object_1_Thread_
+2+^| _Object_2_Thread_
+// row 3
+| Task Thread
+^| _Task_1_1_
+^|__Task_1_2__
+^|_Task_2_1_
+^|_Task_2_2_
+|=====
+
+At some initial stage, all task threads have weakly parallel forward progress
+guarantee. If _K_ is strengthened to have concurrent forward progress
+guarantee, then all of the object threads are also strengthened. Next, in this
+example one task thread for each task sequence is also strengthened. This is
+depicted in the table below (progress guarantee for each thread is in
+parenthesis):
+
+.Possible Progress Guarantees at some time after _WI_ is strengthened.
+[cols="s,,,,"]
+|=====
+// row 1, cells 2 spans 4 cells hence the '4+' before '|'
+| Work Item
+4+^| _WI_ (concurrently parallel)
+// row 2, cells after the first one span 2 cells each
+|Task Sequence Object Thread
+2+^|_Object_1_Thread_ (concurrent)
+2+^| _Object_2_Thread_ (concurrent)
+// row 3
+| Task Thread
+^| _Task_1_1_ (weakly parallel)
+^|__Task_1_2__ (concurrent)
+^|_Task_2_1_ (concurrent)
+^|_Task_2_2_ (weakly parallel)
+|=====
+
+The next example shows how a task thread delegates its progress
+guarantee to another task thread:
+
+Assume that we have a task sequence object _TS_ templated on the function `f`
+with `response_capacity` of 1 and `invocation_capacity` of 5. Four `async` calls
```

Comment:
Does a response_capacity smaller than invocation_capacity make sense?  If you launch more tasks than your response_capacity, none of them are guaranteed to make progress, so how can you be sure the program will ever terminate?

Reply:
I think it could potentially make sense for a scenario where a task sequence may have multiple `async` invocations in flight (i.e. input data gets split into multiple parts, call `task_sequence.async` on each part, input data arrives periodically on some schedule). The task itself takes relatively long to complete, but upon completion, the caller is cleaning them up one after another (for loop of X `get` calls). Kernel needs high async capacity to launch invocations immediately without blocking (so it can go on to do other work), but doesn't expect more async calls to happen until after it has collected all current results. During the period when the kernel has not called `get` and the task has invocation 1 stored in results and invocation 2 ready to store in results, invocation 3 is not strengthened and thus will not make progress. That would be the intended behaviour until the kernel calls get, after which invocation 1 results are retrieved, invocation 2 writes its result and is marked completed, and invocation 3 is strengthened and can make forward progress.